### ISSUE TITLE
Some weird exceptions when building

### ISSUE THREAD

**Issue by wqh17101** [sphinx-doc/sphinx#5306]
Subject: 
Some weird exceptions happened when building
the rtd detailed issues
https://github.com/rtfd/readthedocs.org/issues/4528

### Problem
When i use the RTD to build the doc,i meet an exception
then i build it locally ,but also get the same problem

#### Procedure to reproduce the problem
```
make html
```

#### Error logs / results

![image](https://user-images.githubusercontent.com/26429138/44195149-c2b44700-a16a-11e8-99d4-b7fffe2f91bd.png)
[sphinx-err-nwwq7ojq.log](https://github.com/sphinx-doc/sphinx/files/2292817/sphinx-err-nwwq7ojq.log)

the failed build
https://readthedocs.org/projects/matchzoo/builds/7644468/

#### Expected results
generate normally

### Reproducible project / your project
https://github.com/faneshion/MatchZoo/tree/hotfix/doc

### Environment info
- OS: Linux and Windows
- Python version: python3.6
- Sphinx version: 1.75

**Comment by tk0miya** [comment]
It seems `hotfix/doc` branch was removed. Please let me know how to reproduce the error.

**Comment by wqh17101** [comment]
i found the reason and fixed it . Thank you for your help. There were something wrong for typehints.
https://github.com/faneshion/MatchZoo/pull/290
you can see this @tk0miya

**Comment by tk0miya** [comment]
Thank you for information. I found the error.

This is caused by a bug of utility function of Sphinx: `sphinx.ext.autodoc.inspector:formatargspec()`. But unfortunately, it is not used in sphinx-core because it is deprecated internally. In this case, it is called by `sphinx_autodoc_typehints` extension.

I'll fix the function later.

Note: `typing.get_type_hints` raises an exception if subject has broken type hints:
```
$ cat t.py
import typing

def func(arg: "unknown_type"):
    pass

typing.get_type_hints(func)
$ python -V
Python 3.6.5
$ python t.py
Traceback (most recent call last):
  File "t.py", line 6, in <module>
    typing.get_type_hints(func)
  File "/Users/tkomiya/.pyenv/versions/3.6.5/lib/python3.6/typing.py", line 1535, in get_type_hints
    value = _eval_type(value, globalns, localns)
  File "/Users/tkomiya/.pyenv/versions/3.6.5/lib/python3.6/typing.py", line 350, in _eval_type
    return t._eval_type(globalns, localns)
  File "/Users/tkomiya/.pyenv/versions/3.6.5/lib/python3.6/typing.py", line 245, in _eval_type
    eval(self.__forward_code__, globalns, localns),
  File "<string>", line 1, in <module>
NameError: name 'unknown_type' is not defined
```

Since 1.7, autodoc module uses `sphinx.util.inspect:Signature` instead. In the class, any exceptions from `get_type_hints()` are handled.

**Comment by wqh17101** [comment]
That is great. @tk0miya

**Comment by tk0miya** [comment]
Fixed by #5309.
Thank you for reporting.
